demos/python_demos/formula_recognition_demo/utils.py

```diff
@@ -1,11 +1,340 @@
+                                                                            }
+                                                 )
+
+    def infer_async(self, model_input):
+        model_input = change_layout(model_input)
+        assert self.is_async
+        # asynchronous variant
+        if self.model_status == Model.Status.ready:
+            infer_status_encoder = self._run_encoder(model_input)
+            return None
+
+        if self.model_status == Model.Status.encoder_infer:
+            infer_status_encoder = self._infer_request_handle_encoder.wait(timeout=1)
+            if infer_status_encoder == 0:
+                self._run_decoder()
+            return None
+
+        return self._process_decoding_results()
+
+    def infer_sync(self, model_input):
+        assert not self.is_async
+        model_input = change_layout(model_input)
+        self._run_encoder(model_input)
+        self._run_decoder()
+        res = None
+        while res is None:
+            res = self._process_decoding_results()
+        return res
+
+    def _process_decoding_results(self):
+        timeout = 1 if self.is_async else -1
+        infer_status_decoder = self._infer_request_handle_decoder.wait(timeout)
+        if infer_status_decoder != 0 and self.is_async:
+            return None
+        dec_res = self._infer_request_handle_decoder.output_blobs
+        self._unpack_dec_results(dec_res)
+
+        if self.tgt[0][0][0] == END_TOKEN or self.num_infers_decoder >= self.args.max_formula_len:
+            self.num_infers_decoder = 0
+            self.logits = np.array(self.logits)
+            logits = self.logits.squeeze(axis=1)
+            targets = np.argmax(logits, axis=1)
+            self.model_status = Model.Status.ready
+            return logits, targets
+        self._infer_request_handle_decoder = self._async_infer_decoder(self.row_enc_out,
+                                                                       self.dec_states_c,
+                                                                       self.dec_states_h,
+                                                                       self.output,
+                                                                       self.tgt,
+                                                                       req_id=0
+                                                                       )
+
+        return None
+
+    def _run_encoder(self, model_input):
+        timeout = 1 if self.is_async else -1
+        self._infer_request_handle_encoder = self._async_infer_encoder(model_input, req_id=0)
+        self.model_status = Model.Status.encoder_infer
+        infer_status_encoder = self._infer_request_handle_encoder.wait(timeout=timeout)
+        return infer_status_encoder
+
+    def _run_decoder(self):
+        enc_res = self._infer_request_handle_encoder.output_blobs
+        self._unpack_enc_results(enc_res)
+        self._infer_request_handle_decoder = self._async_infer_decoder(
```

Comment:
This function literally only runs the decoder, the status (`infer_status_decoder = self._infer_request_handle_decoder.wait(timeout)`) is checked in the `_process_decoding_results` function

Reply:
Since now `_run_{en,de}coder()` never wait for results maybe it's better to name them `_start_{en,de}coder()` or `_start_async_infer_{en,de}coder()` to refer to IE's `start_async()`.

Reply:
Renamed to `_start_decoder`, as this function is also used in `infer_sync`, so this would be inconsistent to call it `start_infer_async`